packages/storybook/builder/src/plugin-storybook.ts

```diff
@@ -0,0 +1,383 @@
+    ...(routerPath ? { [`@storybook/router`]: routerPath } : {}),
+    ...(themingPath ? { [`@storybook/theming`]: themingPath } : {}),
+  };
+
+  builderConfig.source ??= {};
+  builderConfig.source.alias = {
+    ...builderConfig.source.alias,
+    ...storybookPaths,
+    ...alias,
+  };
+}
+
+async function applyDefines(builderConfig: AllBuilderConfig, options: Options) {
+  const { presets } = options;
+  const envs = await presets.apply<Record<string, string>>('env');
+
+  builderConfig.source ??= {};
+  builderConfig.source.define = {
+    ...builderConfig.source.define,
+    ...stringifyProcessEnvs(envs),
+    'process.env': JSON.stringify(envs),
+    NODE_ENV: JSON.stringify(process.env.NODE_ENV),
+  };
+}
+
+async function applyHTML(builderConfig: AllBuilderConfig, options: Options) {
+  const {
+    presets,
+    packageJson,
+    configType,
+    features,
+    previewUrl,
+    serverChannelUrl,
+  } = options;
+
+  const [
+    coreOptions,
+    frameworkOptions,
+    logLevel,
+    headHtmlSnippet,
+    bodyHtmlSnippet,
+    template,
+    docsOptions,
+  ] = await Promise.all([
+    presets.apply<CoreConfig>('core'),
+    presets.apply('frameworkOptions'),
+    presets.apply('logLevel', undefined),
+    presets.apply('previewHead'),
+    presets.apply('previewBody'),
+    presets.apply<string>('previewMainTemplate'),
+    presets.apply<DocsOptions>('docs'),
+  ]);
+
+  builderConfig.tools ??= {};
+  builderConfig.tools.htmlPlugin = {
+    ...builderConfig.tools.htmlPlugin,
+    template,
+    filename: 'iframe.html',
+    templateParameters: {
+      ...(builderConfig.tools.htmlPlugin
+        ? // @ts-expect-error
+          builderConfig.tools.htmlPlugin.templateParameters || {}
+        : {}),
+      version: packageJson.version || '',
+      globals: {
+        CONFIG_TYPE: configType,
+        LOGLEVEL: logLevel,
+        FRAMEWORK_OPTIONS: frameworkOptions,
+        CHANNEL_OPTIONS: coreOptions.channelOptions,
+        FEATURES: features,
+        PREVIEW_URL: previewUrl,
+        DOCS_OPTIONS: docsOptions,
+        SERVER_CHANNEL_URL: serverChannelUrl,
+      },
+      headHtmlSnippet,
+      bodyHtmlSnippet,
+    },
+    inject: false,
+  };
+}
+
+async function applyMdxLoader(
+  config: { module?: { rules?: any[] } },
+  options: Options & {
+    mdxPluginOptions?: CompileOptions;
+  },
+) {
+  const { presets, mdxPluginOptions } = options;
+
+  const remarkExternalLinks = await import('remark-external-links');
+  const remarkSlug = await import('remark-slug');
+
+  const mdxLoaderOptions = await presets.apply('mdxLoaderOptions', {
+    skipCsf: true,
+    mdxCompileOptions: {
+      providerImportSource: '@storybook/addon-docs/mdx-react-shim',
+      ...mdxPluginOptions?.mdxCompileOptions,
+      remarkPlugins: [
+        remarkSlug,
+        remarkExternalLinks,
+        ...(mdxPluginOptions?.mdxCompileOptions?.remarkPlugins ?? []),
+      ],
+    },
+  });
+  const mdxLoader = options.features?.legacyMdx1
```

Comment:
Do we need to compat legacyMdx1?

Reply:
It's easy to support, so I added it too

Reply:
Update: we better do so, users can still stay old version stories mdx